warehouse/models/staging/payments/littlepay/stg_littlepay__device_transactions.sql

```diff
@@ -32,13 +32,73 @@ stg_littlepay__device_transactions AS (
         {{ trim_make_empty_string_null('granted_zone_ids') }} AS granted_zone_ids,
         {{ trim_make_empty_string_null('onward_zone_ids') }} AS onward_zone_ids,
-        _line_number,
+        CAST(_line_number AS INTEGER) AS _line_number,
         `instance`,
         extract_filename,
+        {{ extract_littlepay_filename_ts() }} AS littlepay_export_ts,
+        {{ extract_littlepay_filename_date() }} AS littlepay_export_date,
         ts,
-        littlepay_export_ts,
+        -- hash all content not generated by us to enable deduping full dup rows
+        -- hashing at this step will preserve distinction between nulls and empty strings in case that is meaningful upstream
+        {{ dbt_utils.generate_surrogate_key(['participant_id', 'customer_id',
+            'device_transaction_id', 'littlepay_transaction_id', 'device_id', 'device_id_issuer',
+            'type', 'transaction_outcome', 'transction_deny_reason', 'transaction_date_time_utc',
+            'location_id', 'location_scheme', 'location_name', 'zone_id', 'route_id', 'mode',
+            'direction', 'latitude', 'longitude', 'vehicle_id', 'granted_zone_ids',
+            'onward_zone_ids']) }} AS _content_hash,
     FROM source
-    QUALIFY ROW_NUMBER() OVER (PARTITION BY littlepay_transaction_id ORDER BY littlepay_export_ts DESC, transaction_date_time_utc DESC) = 1
+),
+
+add_keys_drop_full_dupes AS (
+    SELECT
+        *,
+        -- generate keys now that input columns have been trimmed & cast and files deduped
+        {{ dbt_utils.generate_surrogate_key(['littlepay_export_ts', '_line_number', 'instance']) }} AS _key,
+        littlepay_transaction_id AS _payments_key,
+    FROM clean_columns
+    {{ qualify_dedupe_full_duplicate_lp_rows() }}
+),
+
+stg_littlepay__device_transactions AS (
+    SELECT
+        participant_id, customer_id,
+            device_transaction_id,
+            littlepay_transaction_id,
+            device_id,
+            device_id_issuer,
+            type,
+            transaction_outcome,
+            transction_deny_reason,
+            transaction_date_time_utc,
+            transaction_date_time_pacific,
+            location_id,
+            location_scheme,
+            location_name,
+            zone_id,
+            route_id,
+            mode,
+            direction,
+            latitude,
+            longitude,
+            vehicle_id,
+            granted_zone_ids,
+            onward_zone_ids,
+            _line_number,
+            `instance`,
+            extract_filename,
+            littlepay_export_ts,
+            littlepay_export_date,
+            ts,
+            _key,
+            _payments_key,
+            _content_hash,
+    FROM add_keys_drop_full_dupes
+    -- Some transactions have placeholder information for routes, stops, and directions in their first export,
+    -- then a later export contains the canonical version of the transaction with that information corrected
+    QUALIFY ROW_NUMBER() OVER (
```

Comment:
Does this persist the 4 duplicate rows we talked about where there are two transaction IDs for otherwise identical info? Should those be dropped?

Reply:
I didn't want to add logic that would drop those until we knew more about what we should be seeing. Especially in light of your notes from the recent meeting with Littlepay, it seems like it might be premature to write logic that collapse on shared values in other fields when our current docs tell us to treat `littlepay_transaction_id` as unique.

Reply:
Yeah I can see that, I do wonder if ultimately some handling akin to authorisations where rows are dropped later if they don't map to something could be appropriate? Should we at least write an issue to document that this is being left in place?

Reply:
Confirmed offline that @SorenSpicknall will write an issue after PR merges.